src/pybind/mgr/cephadm/module.py

```diff
@@ -2431,8 +2431,8 @@ def _calc_daemon_deps(self,
             # this way we force a redeploy after a mgr failover
             deps.append(self.get_active_mgr().name())
             deps.append(str(self.get_module_option_ex('prometheus', 'server_port', 9283)))
-            deps += [s for s in ['node-exporter', 'alertmanager', 'ingress']
-                     if self.cache.get_daemons_by_service(s)]
+            deps.append(str(self.service_discovery_port))
+            deps += [s for s in ['node-exporter', 'alertmanager', 'ingress'] if self.cache.get_daemons_by_service(s)]
```

Comment:
found something interesting while testing this. Added a log statement to print out the old and new deps during `_check_daemons` for each daemon to see this working. For changing the service discovery port, it all worked as intended.
```
ceph config set mgr mgr/cephadm/service_discovery_port 8494
```
caused
```
2022-10-06T01:17:06.057476+0000 mgr.vm-00.rksyiw (mgr.34352) 854 : cephadm [INF] Changing service discovery port from 7159 to 8492...
2022-10-06T01:17:06.057618+0000 mgr.vm-00.rksyiw (mgr.34352) 855 : cephadm [INF] [06/Oct/2022:01:17:06] ENGINE Bus STOPPING
2022-10-06T01:17:06.076262+0000 mgr.vm-00.rksyiw (mgr.34352) 856 : cephadm [INF] [06/Oct/2022:01:17:06] ENGINE HTTP Server cherrypy._cpwsgi_server.CPWSGIServer(('192.168.122.54', 7159)) shut down
2022-10-06T01:17:06.094348+0000 mgr.vm-00.rksyiw (mgr.34352) 857 : cephadm [INF] [06/Oct/2022:01:17:06] ENGINE HTTP Server cherrypy._cpwsgi_server.CPWSGIServer(('192.168.122.54', 7150)) shut down
2022-10-06T01:17:06.094791+0000 mgr.vm-00.rksyiw (mgr.34352) 858 : cephadm [INF] [06/Oct/2022:01:17:06] ENGINE Bus STOPPED
2022-10-06T01:17:07.374859+0000 mgr.vm-00.rksyiw (mgr.34352) 859 : cephadm [INF] [06/Oct/2022:01:17:07] ENGINE Bus STARTING
2022-10-06T01:17:07.489455+0000 mgr.vm-00.rksyiw (mgr.34352) 861 : cephadm [INF] [06/Oct/2022:01:17:07] ENGINE Serving on https://192.168.122.54:7150
2022-10-06T01:17:07.618191+0000 mgr.vm-00.rksyiw (mgr.34352) 862 : cephadm [INF] [06/Oct/2022:01:17:07] ENGINE Serving on https://192.168.122.54:8492
2022-10-06T01:17:07.619792+0000 mgr.vm-00.rksyiw (mgr.34352) 863 : cephadm [INF] [06/Oct/2022:01:17:07] ENGINE Bus STARTED
```
and
```
2022-10-06T01:17:07.783871+0000 mgr.vm-00.rksyiw (mgr.34352) 872 : cephadm [INF] Reconfiguring prometheus.vm-01 (dependencies changed)...
2022-10-06T01:17:07.787574+0000 mgr.vm-00.rksyiw (mgr.34352) 873 : cephadm [INF] Reconfiguring daemon prometheus.vm-01 on vm-01
```
which is great, working as intended. However, I noticed in my new added log statement for the prometheus deps
```
2022-10-06T01:17:07.783808+0000 mgr.vm-00.rksyiw (mgr.34352) 871 : cephadm [ERR] daemon: prometheus.vm-01 | new deps: ['8492', '9283', 'alertmanager', 'mgr.vm-00.rksyiw', 'node-exporter'] | old deps: ['7159', '9283', 'alertmanager', 'mgr.vm-00.rksyiw', 'node-exporter']
```
as you can see, some of the deps are just `alertmanager` and `node-exporter`. Where as you look at the same log statement for grafana
```
2022-10-06T01:18:16.168248+0000 mgr.vm-00.rksyiw (mgr.34352) 928 : cephadm [ERR] daemon: grafana.vm-00 | new deps: ['prometheus.vm-01'] | old deps: ['prometheus.vm-01']
```
and its dep is the actual full prometheus daemon name, as it should be. I think https://github.com/ceph/ceph/commit/a0fe3ffdaaddd661f115de78ddae89b07b98a9ae introduced a regression where the prometheus deps are just getting the service type ("alertmanager", "node-exporter" or "ingress") as deps instead of the actual daemons in the service. Some further testing confirmed that we don't actually reconfig prometheus anymore if node-exporter daemons are moved to a different set of hosts.

Technically, this regression isn't caused by this PR, so I thought I'd leave it up to you if you want to fix it in this PR or in a follow up. If you want it as a follow up we can probably merge this as I think everything else is working fine. Let me know what you think.

Reply:
Thank you very much for the detailed testing and for the feedback. The goal from the PR https://github.com/ceph/ceph/pull/46400 was basically to make the whole prometheus configuration dynamic. That means we don't depend on a _static_ configuration anymore, instead, we just get the prometheus config (scrape and alert configs) dynamically by using the new service-discovery mechanism. So right now, our generated `prometheus.yml` looks like:

```
.....
scrape_configs:
  - job_name: 'ceph'
    honor_labels: true
    http_sd_configs:
    - url: https://192.168.100.100:8765/sd/prometheus/sd-config?service=mgr-prometheus
      tls_config:
        ca_file: root_cert.pem

  - job_name: 'node'
    http_sd_configs:
    - url: https://192.168.100.100:8765/sd/prometheus/sd-config?service=node-exporter
      tls_config:
        ca_file: root_cert.pem
```
So for example for node-exporter prometheus get the nodes list from the url `https://192.168.100.100:8765/sd/prometheus/sd-config?service=node-exporter`. So, if we curl this url we get:

```
[
  {
    "targets": [
      "192.168.100.100:9100"
    ],
    "labels": {
      "instance": "ceph-node-0"
    }
  },
  {
    "targets": [
      "192.168.100.101:9100"
    ],
    "labels": {
      "instance": "ceph-node-1"
    }
  },
  {
    "targets": [
      "192.168.100.102:9100"
    ],
    "labels": {
      "instance": "ceph-node-2"
    }
  }
]

```
The thing is, I think with this new dynamic configuration we don't have to adjust the promteheus deps to depend on the daemons names anymore (maybe we can also remove the service-type! as it's in fact useless) as we just get the conf dynamically. If you move some node-exporter daemon to some other host we should just get the new list from the above url updated and that's whay prometheus will be using. 

WDYT?

PD: Regardless of whether we need an additional PR to fix this dep issues or not I also think it's better idea to merge this PR and then open a new one to address any deps related issues.

Reply:
okay, in that case we should definitely just remove the deps. Having the service names in there isn't very useful. I'm okay with this being done in a separate PR